packages/apollo/src/commands/service/check.ts

```diff
@@ -253,112 +260,198 @@ export default class ServiceCheck extends ProjectCommand {
+                  | { schema: IntrospectionSchemaInput }
+                  | undefined;
 
-                  const historicParameters = validateHistoricParams({
-                    validationPeriod: flags.validationPeriod,
-                    queryCountThreshold: flags.queryCountThreshold,
-                    queryCountThresholdPercentage:
-                      flags.queryCountThresholdPercentage
-                  });
+                // If we're `federated`, then run composition validation. When we're using composition
+                // validation we'll receive a schema has that represents the composed schema.
+                if (ctx.federationSchemaHash) {
+                  schemaCheckSchemaVariables = {
+                    schemaHash: ctx.federationSchemaHash
+                  };
+                } else {
+                  // This is _not_ a `federated` schema. Resolve the schema given `config.tag`.
+                  task.output = "Resolving schema";
+                  schema = await project.resolveSchema({ tag: config.tag });
+                  if (!schema) {
+                    throw new Error("Failed to resolve schema");
+                  }
 
-                  task.output = "Validating schema";
-                  const variables: CheckSchemaVariables = {
-                    id: configName,
-                    // @ts-ignore
-                    // XXX Looks like TS should be generating ReadonlyArrays instead
-                    schema: introspectionFromSchema(schema).__schema,
-                    tag: flags.tag,
-                    gitContext: await gitInfo(this.log),
-                    frontend: flags.frontend || config.engine.frontend,
-                    ...(historicParameters && { historicParameters })
+                  schemaCheckSchemaVariables = {
+                    schema: introspectionFromSchema(schema)
+                      .__schema as IntrospectionSchemaInput
                   };
-                  const { schema: _, ...restVariables } = variables;
-                  this.debug("Variables sent to Engine:");
-                  this.debug(restVariables);
+                }
+
+                await gitInfo(this.log);
+
+                const historicParameters = validateHistoricParams({
+                  validationPeriod: flags.validationPeriod,
+                  queryCountThreshold: flags.queryCountThreshold,
+                  queryCountThresholdPercentage:
+                    flags.queryCountThresholdPercentage
+                });
+
+                task.output = "Validating schema";
+
+                const variables: CheckSchemaVariables = {
+                  id: configName,
+                  tag: flags.tag,
+                  gitContext: await gitInfo(this.log),
+                  frontend: flags.frontend || config.engine.frontend,
+                  ...(historicParameters && { historicParameters }),
+                  // `variables` will either contain `schemaHash` or `schema`, not both.
```

Comment:
Is this an invariant we can enforce/pretty log in the client side?

Reply:
We're enforcing it above, this comment is misplaced. I'm going to remove it.